src/snowflake/snowpark/session.py

```diff
@@ -680,14 +681,30 @@ def _resolve_packages(
         validate_package: bool = True,
         include_pandas: bool = False,
     ) -> List[str]:
+        package_names = list()
+        for package in packages:
+            if isinstance(package, ModuleType):
+                package_name = MODULE_NAME_TO_PACKAGE_NAME_MAP.get(
+                    package.__name__, package.__name__
+                )
+                package = f"{package_name}=={pkg_resources.get_distribution(package_name).version}"
+            else:
+                package = package.strip().lower()
+            package_names.append(pkg_resources.Requirement.parse(package).key)
```

Comment:
You can save (package name -> (package requirement identifier, is_from_local)) as a dict, which can be used later in this function. Then you can avoid duplication here https://github.com/snowflakedb/snowpark-python/blob/6f0616627e53780ceb0ab897d92e60fad1365c91/src/snowflake/snowpark/session.py#L698-L709.